kernel/aster-nix/src/taskless.rs

```diff
@@ -0,0 +1,220 @@
+///
+/// ```
+/// Note that the callback function should not involve sleep instructions.
+pub struct Taskless {
+    /// Whether the taskless has been scheduled in a taskless list.
+    is_scheduled: AtomicBool,
+    /// Whether the taskless is running.
+    is_running: AtomicBool,
+    /// The function that will be called when execute this taskless.
+    callback: Box<RefCell<dyn FnMut() + Send + Sync + 'static>>,
+    /// If this taskless is disabled.
+    is_disabled: AtomicBool,
+    link: LinkedListAtomicLink,
+}
+
+intrusive_adapter!(TasklessAdapter = Arc<Taskless>: Taskless { link: LinkedListAtomicLink });
+
+cpu_local! {
+    static TASKLESS_LIST: SpinLock<LinkedList<TasklessAdapter>> = SpinLock::new(LinkedList::new(TasklessAdapter::NEW));
+    static TASKLESS_HI_LIST: SpinLock<LinkedList<TasklessAdapter>> = SpinLock::new(LinkedList::new(TasklessAdapter::NEW));
+}
+
+impl Taskless {
+    /// Create a new `Taskless` instance with its callback function.
+    pub fn new<F>(callback: F) -> Arc<Self>
+    where
+        F: FnMut() + Send + Sync + 'static,
+    {
+        // Since the same taskless will not be executed concurrently,
+        // it is safe to use a `RefCell` here though the `Taskless` will
+        // be put into an `Arc`.
+        #[allow(clippy::arc_with_non_send_sync)]
+        Arc::new(Self {
+            is_scheduled: AtomicBool::new(false),
+            is_running: AtomicBool::new(false),
+            callback: Box::new(RefCell::new(callback)),
+            is_disabled: AtomicBool::new(false),
+            link: LinkedListAtomicLink::new(),
+        })
+    }
+
+    /// Schedule the taskless.
+    ///
+    /// This operation will put the taskless into normal taskless list,
+    /// and it will be executed soon.
+    ///
+    /// If the taskless has been scheduled, this function will do nothing.
+    pub fn schedule(self: &Arc<Self>) {
+        do_schedule(self, &TASKLESS_LIST);
+        SoftIrqLine::get(TASKLESS_SOFTIRQ_ID).raise();
+    }
+
+    /// Schedule the taskless with high priority.
+    ///
+    /// This operation will put the taskless into high priority taskless list,
+    /// and it will be executed soon.
+    ///
+    /// If the taskless has been scheduled, this function will do nothing.
+    pub fn schedule_with_hi_priority(self: &Arc<Self>) {
+        do_schedule(self, &TASKLESS_HI_LIST);
+        SoftIrqLine::get(TASKLESS_HI_SOFTIRQ_ID).raise();
+    }
+
+    /// Enable this taskless.
+    pub fn enable(self: &Arc<Self>) {
+        self.is_disabled.store(false, Ordering::Release);
+    }
+
+    /// Disable this taskless.
+    pub fn disable(self: &Arc<Self>) {
+        self.is_disabled.store(true, Ordering::Release);
+    }
+}
+
+fn do_schedule(
+    taskless: &Arc<Taskless>,
+    taskless_list: &'static CpuLocal<SpinLock<LinkedList<TasklessAdapter>>>,
+) {
+    if taskless.is_scheduled.load(Ordering::Acquire) {
+        return;
+    }
+
+    taskless.is_scheduled.store(true, Ordering::Release);
```

Comment:
The `do_schedule` function is vulnerable to race conditions. Consider the following concurrent execution sequences.

```
Thread A                        |   Thread B
is_scheduled.load() == false    |
                                |   is_scheduled.load() == false
is_scheduled.store(true)        |
push_front(taskless.clone()     |
                                |   is_scheduled.store(true)
                                |   push_front(taskless.clone()
```

Given that fact that this simple `do_schedule` function is vulnerable to race conditions, I am worried about the correctness of `taskless_softirq_handler` under concurrency.

Please examine `taskless_softirq_handler` and `do_schedule` more carefully for potential race conditions.